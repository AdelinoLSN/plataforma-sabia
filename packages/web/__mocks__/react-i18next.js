module.exports = {
	useTranslation: jest.fn(() => {
		return {
			t(slug) {
				return slug;
			},
		};
	}),
};
