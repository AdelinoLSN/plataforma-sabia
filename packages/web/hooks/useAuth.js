import { useContext } from 'react';
import UserContext from '../components/User/UserContext';

function useAuth() {
	return useContext(UserContext);
}

export default useAuth;
