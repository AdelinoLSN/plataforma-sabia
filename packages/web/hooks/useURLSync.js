import { useState, useEffect } from 'react';
import qs from 'query-string';

const DEBOUNCE_TIME = 200;

const encodedCategories = {
	agua: 'Água',
	saneamento: 'Saneamento',
	'energia-eletrica': 'Energia Elétrica',
	'energia-solar': 'Energia Solar',
};

const decodedCategories = Object.keys(encodedCategories).reduce((acc, key) => {
	const newKey = encodedCategories[key];
	const newValue = key;

	return {
		...acc,
		[newKey]: newValue,
	};
}, {});

const createURL = (state) => {
	const isDefaultRoute =
		!state.query &&
		state.page === 1 &&
		state.refinementList &&
		state.refinementList.category.length === 0;

	if (isDefaultRoute) {
		return '/search';
	}

	const queryParameters = {};

	if (state.query) {
		queryParameters.query = encodeURIComponent(state.query);
	}
	if (state.page !== 1) {
		queryParameters.page = state.page;
	}
	if (state.refinementList.category) {
		queryParameters.categories = state.refinementList.category
			.map((category) => decodedCategories[category] || category)
			.map(encodeURIComponent);
	}

	const queryString = qs.stringifyUrl(
		{ url: '', query: queryParameters },
		{ arrayFormat: 'comma' },
	);

	return `/search${queryString}`;
};

const searchStateToURL = (searchState) => (searchState ? createURL(searchState) : '');

const urlToSearchState = (location) => {
	const url = qs.parse(location.search.slice(1));
	const { query = '', page = 1 } = url;
	const allCategories = url?.categories ? url.categories.split(',') : [];

	return {
		query: decodeURIComponent(query),
		page,
		refinementList: {
			category: allCategories
				.map((category) => encodedCategories[category] || category)
				.map(decodeURIComponent),
		},
	};
};

function useURLSync() {
	const [searchState, setSearchState] = useState({});
	const [debouncedSetState, setDebouncedSetState] = useState(null);

	const onPopState = ({ state }) => {
		setSearchState(state || {});
	};

	useEffect(() => {
		setSearchState(urlToSearchState(window.location));
		window.addEventListener('popstate', onPopState);
		return () => {
			window.removeEventListener('popstate', onPopState);
		};
	}, []);

	const onSearchStateChange = (newSearchState) => {
		clearTimeout(debouncedSetState);

		setDebouncedSetState(
			setTimeout(() => {
				window.history.pushState(newSearchState, null, searchStateToURL(newSearchState));
			}, DEBOUNCE_TIME),
		);

		setSearchState(newSearchState);
	};

	return { searchState, createURL, onSearchStateChange };
}

export default useURLSync;
