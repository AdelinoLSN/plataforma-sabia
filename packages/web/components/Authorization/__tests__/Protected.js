import React from 'react';
import { useTranslation } from 'react-i18next';
import { render } from 'test-utils';
import { Protected } from '..';
import { useModal } from '../../../hooks';

jest.mock('../../../hooks', () => {
	return {
		__esModule: false,
		useModal: () => {
			return {
				openModal: jest.fn(),
				closeModal: jest.fn(),
			};
		},
		useAuth: () => {
			return {
				user: jest.fn(),
			};
		},
	};
});

test('it should not render children if user is not logged in', () => {
	const { t } = useTranslation(['common']);

	const { container } = render(
		<Protected>
			<h1>Children</h1>
		</Protected>,
	);

	expect(container).toMatchSnapshot();

	const { openModal } = useModal();

	expect(openModal).toHaveBeenCalledWith('login', {
		message: t('common:signInToContinue'),
	});
});
