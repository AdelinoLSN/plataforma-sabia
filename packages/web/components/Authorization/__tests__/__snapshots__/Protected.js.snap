// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should not render children if user is not logged in 1`] = `<div />`;
